examples/step-77/step-77.cc

```diff
@@ -554,77 +557,148 @@ namespace Step77
+                  const Vector<double> & /*current_f*/) {
+                compute_and_factorize_jacobian(current_u);
+
+                return 0;
+              };
+
+            nonlinear_solver.solve_with_jacobian =
+              [&](const Vector<double> &rhs,
+                  Vector<double> &      dst,
+                  const double          tolerance) {
+                this->solve(rhs, dst, tolerance);
+
+                return 0;
+              };
+
+            nonlinear_solver.solve(current_solution);
+          }
+        else
+#endif
+#ifdef DEAL_II_WITH_TRILINOS
+          if (non_liner_solver_name == "NOX")
+          {
+            // 1) configure solver
+            TrilinosWrappers::NOXSolver<Vector<double>>::AdditionalData
+              additional_data(1000, target_tolerance, 1e-8);
+
+            Teuchos::RCP<Teuchos::ParameterList> non_linear_parameters =
+              Teuchos::rcp(new Teuchos::ParameterList);
+
+            non_linear_parameters->set("Nonlinear Solver", "Line Search Based");
+
+            auto &print_params = non_linear_parameters->sublist("Printing");
+            print_params.set("Output Information", 0);
+
+            auto &dir_parameters = non_linear_parameters->sublist("Direction");
+            dir_parameters.set("Method", "Newton");
+
+            auto &search_parameters =
+              non_linear_parameters->sublist("Line Search");
+            search_parameters.set("Method", "Polynomial");
+
+            // 2) set up solver
+            TrilinosWrappers::NOXSolver<Vector<double>> nonlinear_solver(
+              additional_data, non_linear_parameters);
+
+            nonlinear_solver.residual =
+              [&](const Vector<double> &evaluation_point,
+                  Vector<double> &      residual) {
+                compute_residual(evaluation_point, residual);
+
+                return 0;
+              };
+
+            nonlinear_solver.setup_jacobian =
+              [&](const Vector<double> &current_u) {
+                compute_and_factorize_jacobian(current_u);
+
+                return 0;
+              };
+
+            nonlinear_solver.apply_jacobian = [&](const Vector<double> &src,
+                                                  Vector<double> &      dst) {
+              this->jacobian_matrix.vmult(dst, src);
 
               return 0;
             };
 
-          nonlinear_solver.solve_with_jacobian = [&](const Vector<double> &rhs,
-                                                     Vector<double> &      dst,
-                                                     const double tolerance) {
-            this->solve(rhs, dst, tolerance);
+            nonlinear_solver.solve_with_jacobian =
+              [&](const Vector<double> &rhs,
+                  Vector<double> &      dst,
+                  const double          tolerance) {
+                this->solve(rhs, dst, tolerance);
 
-            return 0;
-          };
+                return 0;
+              };
```

Comment:
What if you declared these lambda functions in common code above the switch between solvers, and then here just did the assignment? That would minimize the amount of code that actually deviates between the two choices.

Reply:
> What if you declared these lambda functions in common code above the switch between solvers, and then here just did the assignment?

I have moved the lamdba functions. However, there are some differences: the NOX wrapper does need a function to intialize the vector and for setting up the Jacobian the residuum is not available. Also the NOX wrappers have additional lambda functions `apply_jacobian`, `check_iteration_status`, `setup_preconditioner`, and `update_preconditioner_predicate`. 

Latter two allow to setup the preconditoner less frequently and are additions by me, since the NOX interface does not have a notion for preconditioning however the official wrappers for Epetra have it... This is quite annoying and indicate where the focus of the project is... 

Furthermore, it seems that `NOX` updates the `Jacobian` after each Newton step in contrast to KINSOL. For me this is not a problem since I normally don't assemble the matrix. But I one might want to add such a feature sometime in the future so that we mimic what KINSOL does.